<a href="https://colab.research.google.com/github/hrittik2002/ML-Notes/blob/main/python/iterator_and_generator.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Python Iterators

## 1.1. What is Iteration

Iteration is a general term for taking each item of something, one after another. Any time you use a loop, explicit or implicit, to go over a group of items, that is iteration.

In [1]:
# Example
num = [1,2,3]

for i in num:
  print(i)

1
2
3


## 1.2. What is Iterator

- An iterator is a object that allows the programmer to traverse through a sequence of data **without having to store the entire data in the memory**.

- Iterator is that object due to which iteration is possible. Interally when we are running a loop, there is an iterator object that fetch the items one by one and give it to loop.

- The Best thing about Iterator is it loads one item in the memory at a time, then perform all the operations then remove that item from the memory, then loads the next item in the memory. Due to this it is much memory efficient.

- Every iterator contains **an iter function and a next function**.

In [4]:
# Lets say we have to print the multiplication by 2 of all the first 100 natural numbers.
import sys

# without iterator
L = [x for x in range(1,100)]

# for i in L:
#   print(i*2)

# how much memory occupied
sys.getsizeof(L) # bytes

920

In [5]:
# with iterator
x = range(1,100)

# for i in x:
#   print(i*2)

# how much memory occupied
sys.getsizeof(x) # bytes

48

Here,
- when we are using list L, here in the memory we are loading all the items at once, so it is taking 920 bytes memory.
- In the secound case, we are using range (internally iterator), here 1 item is loading the memory at a time so it is taking 48 bytes memory.  

#  1.3. What is Iterable

- Iterable is an object, which one can iterate over. Means any obeject in Python in which we can run loop and get items of that object one by one then it is iterable.

- Note for iterator the items loads in the memory 1 at a time. But here it is not the case, any item where we can apply loop is iterable.

- Means list, dictionary, set, tuple, range all of the 5are iterable. But it we talk about iterator only range is iterator.

- Every iterator is iterable. But all iterable are not iterator.

- Iterable Object generates an iterator when passed to i**ter() methods**.

- When we run loop over a iterable interally it uses the iter() method to generate iterator.

In [14]:
# Example

L = [1,2,3]

print(type(L))
# Here List L is a iterable

print(type(iter(L)))
# Here List L is a iterator

<class 'list'>
<class 'list_iterator'>


## 1.4. iter() and next() function

- Every iterable contains a iter() method

In [12]:
# Example

L = [1,2,3]

print(type(L)) # Here List L is a iterable

dir(L) # To verify check L must contain __iter__ method

<class 'list'>


['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

- Every iterator contains a iter function and a next function

In [13]:
# Example

L = [1,2,3]

print(type(L))
# Here List L is a iterable

iterator_obj = iter(L) # we can generate iterator object using __iter__

print(type(iterator_obj))

dir(iterator_obj) # To verify check iterator_obj must contain __iter__ and __next__

<class 'list'>
<class 'list_iterator'>


['__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__length_hint__',
 '__lt__',
 '__ne__',
 '__new__',
 '__next__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__']

- If we pass a iterable object through iter() then we get a iterator object.

- If we pass a iterator object through iter() then we get the same iterator object.

- If we pass a iterator object through next() then we get the next item.

In [24]:
# Example
L = [1,2,3] # List is a iterable

 # We can generate a iterator when we pass iterable to __iter__ function
iterator_obj = iter(L)
print(iterator_obj)

# If we again pass the iterator object to __iter__ function
print(iter(iterator_obj)) # check the output carefully we are getting the same iterator object

In [25]:
# we use __next__ function to get the next item
print(next(iterator_obj)) # we get 1
print(next(iterator_obj)) # we get 2
print(next(iterator_obj)) # we get 3
print(next(iterator_obj)) # we get Exception StopIteration

1
2
3


StopIteration: 

## 1.5. How for loop works?


In [26]:
num = [1,2,3]

for i in num:
  print(i)

1
2
3


In [27]:
# interanlly
num = [1,2,3]

# Step-1 : Fetch the iterator
iter_num = iter(num)

# Step-2 : next
print(next(iter_num))
print(next(iter_num))
print(next(iter_num))
print(next(iter_num)) # running __next__ till we getting StopIteration

1
2
3


StopIteration: 

Lets create a custom for loop

In [29]:
def custom_for_loop(iterable):
  iterator = iter(iterable) # Step-1 : generate iterator
  while True:
    try:
      print(next(iterator)) # Step-2 : get next item till exception
    except StopIteration:
      break

# test it
a = [1,2,3]
custom_for_loop(a)

1
2
3


## 1.6 create custom range() function

In [30]:
class custom_range: # iterable

  def __init__(self, start, end):
    self.start = start
    self.end = end

  def __iter__(self):
    return custom_range_iterator(self) # return iterator object


class custom_range_iterator: # iterator

  def __init__(self,iterable_obj):
    self.iterable = iterable_obj

  def __iter__(self):
    return self # when we run __iter__ over iterator it return self

  def __next__(self):
    if self.iterable.start >= self.iterable.end:
      raise StopIteration
    current = self.iterable.start
    self.iterable.start += 1
    return current

In [31]:
for i in custom_range(1,10):
  print(i)

1
2
3
4
5
6
7
8
9
